### PR TITLE
[CARBONDATA-1804] Support Plug-gable File Operations based on File types

### PR DESCRIPTION
**Pull request by ManoharVanam**
**Description :** 
Refactor FileFactory based on FileType to support plug-gable file handlers so that custom file handlers can have their specific logic.
Example : User can provide his own implementations by extending existing FileTypes
**Refactore FileFactory code** : Moved File type opearation code into specific file types instead of checking file type every time in FileFactory. So that custom file operations can be performed on the specific file types by extending required File type

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
Not, Old test cases will take care
        - How it is tested? Please attach test report.
Verified in cluster
        - Is it a performance related change? Please attach the performance test report.
NO
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by ManoharVanam** [comment]
retest this please

**Comment by ravipesala** [comment]
retest this please

**Comment by ravipesala** [comment]
retest sdv please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1431/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1858/

**Comment by jackylk** [comment]
Please provide description of this PR @ManoharVanam

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/DFSFileHolderImpl.java [hunk 1]
```diff
@@ -35,7 +35,7 @@
   /**
    * cache to hold filename and its stream
    */
-  private Map<String, FSDataInputStream> fileNameAndStreamCache;
+  public Map<String, FSDataInputStream> fileNameAndStreamCache;
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/datastore/impl/DFSFileHolderImpl.java]: member variable should not be public

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala [hunk 2]
```diff
@@ -353,14 +355,14 @@ class CarbonFileMetastore extends CarbonMetaStore {
...
     tableInfo.setMetaDataFilepath(schemaMetadataPath)
-    val fileType = FileFactory.getFileType(schemaMetadataPath)
-    if (!FileFactory.isFileExist(schemaMetadataPath, fileType)) {
-      FileFactory.mkdirs(schemaMetadataPath, fileType)
-    }
-    val thriftWriter = new ThriftWriter(schemaFilePath, false)
-    thriftWriter.open(FileWriteOperation.OVERWRITE)
-    thriftWriter.write(thriftTableInfo)
-    thriftWriter.close()
+      val fileType = FileFactory.getFileType(schemaMetadataPath)
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala]: It seems indentation not correct

Comment (ManoharVanam) [reply integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala]: ok

File: core/src/main/java/org/apache/carbondata/core/datastore/impl/DFSFileHolderImpl.java (hunk 1 above)
Comment (ManoharVanam) [reply core/src/main/java/org/apache/carbondata/core/datastore/impl/DFSFileHolderImpl.java]: ok

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1456/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1874/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1459/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1876/

**Comment by ManoharVanam** [comment]
retest this please

**Comment by manishgupta88** [comment]
LGMT

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1500/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1542/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1943/

**Comment by jackylk** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1553/

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**